command/harness/setup.go

```diff
@@ -97,30 +94,43 @@ func HandleSetup(ctx context.Context, r *SetupVMRequest, s store.StageOwnerStore
 		return nil, fmt.Errorf("pool not defined")
 	}
 
+	cleanUpFn := func() {
+		errCleanUpStageOwner := s.Delete(context.Background(), stageRuntimeID)
+		if errCleanUpStageOwner != nil {
+			logr.WithError(errCleanUpStageOwner).Errorln("failed to delete failed stage owner")
+		}
+	}
+
+	if err := s.Create(ctx, &types.StageOwner{StageID: stageRuntimeID, PoolName: pool}); err != nil {
+		return nil, fmt.Errorf("could not create stage owner entity: %w", err)
+	}
+
 	instance, err := poolManager.Provision(ctx, pool, env.Runner.Name, env.Settings.LiteEnginePath)
 	if err != nil {
+		go cleanUpFn()
 		return nil, fmt.Errorf("failed provisioning")
 	}
 
-	if instance.IsHibernated {
-		instance, err = poolManager.StartInstance(ctx, pool, instance.ID)
-		if err != nil {
-			return nil, fmt.Errorf("failed to start the instance up")
-		}
-	}
-
 	logr = logr.
 		WithField("ip", instance.Address).
 		WithField("id", instance.ID)
 
 	// cleanUpFn is a function to terminate the instance if an error occurs later in the handleSetup function
-	cleanUpFn := func() {
+	cleanUpFn = func() {
 		errCleanUp := poolManager.Destroy(context.Background(), pool, instance.ID)
```

Comment:
should we add the cleanup for the db mapping here as well? eg what happens if call fails later - the mapping would still remain right?

In the logs, I saw one of the executions failed with the error on L152. If that happens, we still want to remove the mapping so that it successfully retries right?

Reply:
It does happen on the harness destroy. The reason it’s not been added at the manager level is because the table isn’t used by drone. It’s impl is unique to delegate. I do think maybe we should add it at manager level but it’s a bigger refactor with higher chance of regression